### ISSUE TITLE
Replies are not going to the most recent message that involves you - whether it is received or sent. (Former behavior.)

### ISSUE THREAD

**Issue by mibby** [EssentialsX/Essentials#250]
Essentials dev 216.

Replies are not going to the most recent message that involves you - whether they are sent by you to other players or received from other players. (Former behavior.) Instead, it goes to the most recent message from a player that you received. 

```
last-message-reply-recipient: false
```

Scenario.

```
Mibby messages Bob.
/msg bob hi
(Bob receives message.)

Bob replies.
/r hi
(Mibby receives message.)

Mibby replies.
/r how are you?
(Bob receives message.)

Mibby messages Tom.
/msg tom hi
(Tom receives message.)

Mibby repies.
/r how are you?
(Bob receives message.)
[PROBLEM]
```

Or in this case, players can't reply to a message they receive.

```
last-message-reply-recipient: true
```

Scenario.

```
Mibby messages Bob.
/msg bob hi
(Bob receives message.)

Bob replies.
/r hi
(No message is received - bob receives "You have nobody to whom you can reply.")
[PROBLEM]
```

Is there no way to go back to the old behavior where replies go to the last message that involves you?

**Comment by mibby** [comment]
Tagging relevant parties.

@SupaHam @drtshock @vemacs

**Comment by drtshock** [comment]
We watch the repo so we get notifications anyway. Tagging is unnecessary.

**Comment by SupaHam** [comment]
Hi, thank you for your report. I've investigated the given scenarios and have found them invalid.

When `last-message-reply-recipient` is `true`, the new behaviour is put in place. However, when When `last-message-reply-recipient` is `false`, the old behaviour is put in place.

When `true`, your _reply-recipient_ becomes whoever **YOU** messaged last, not whoever messaged **YOU** last. Otherwise, when `false`, your _reply-recipient_ becomes whoever messaged **YOU** last, not whoever **YOU** messaged last. 

**Edit:** To go back to the old behaviour, where your _reply-recipient_ is whoever messaged you last, set `last-message-reply-recipient` to `false`.

**Comment by mibby** [comment]
@SupaHam I have `last-message-reply-recipient` set to `false` but I run into this problem.

```
Mibby messages Bob.
/msg bob hi
(Bob receives message.)

Bob replies.
/r hi
(Mibby receives message.)

Mibby replies.
/r how are you?
(Bob receives message.)

Mibby messages Tom.
/msg tom hi
(Tom receives message.)

Mibby repies.
/r how are you?
(Bob receives message.)
[PROBLEM]
```

In otherwords, my replies go to the last person who messaged me, not who I messaged. I'm trying to go back to the old behavior where /r goes to the last message you were a part of, whether it was something you received from another player or a message you sent to someone else.

**Comment by SupaHam** [comment]
Yes, that is expected behaviour. As I said above

> Edit: To go back to the old behaviour, where your reply-recipient is whoever messaged you last, set last-message-reply-recipient to false.

In your case, you want to use the new behaviour, where your replies go to the last person you messaged. So, you need to turn `last-message-reply-recipient` to `true`.

**Comment by mibby** [comment]
With `last-message-reply-recipient` set to `true`,  players who had not previously used /msg can't quickly reply to messages they received since it states they have no one to reply to. Again, in this scenario.

```
Mibby messages Bob.
/msg bob hi
(Bob receives message.)

Bob replies.
/r hi
(No message is received - bob receives "You have nobody to whom you can reply.")
[PROBLEM]
```

**Comment by SupaHam** [comment]
When did you test this functionality? I've tested it just yesterday and it worked just fine. I haven't changed anything since.

**Comment by mibby** [comment]
I tested it a few hours ago between alternate accounts on dev build 216.

**Comment by SupaHam** [comment]
I just tested it as well :s 
http://supa.me/vE36fu.png

**Comment by mibby** [comment]
Edit: Are you messaging yourself? I am testing with one player messaging another account, then that other account using /r to reply without first /msging them back.

Edit2: Another example of why it doesn't work and isn't the old behavior...

Scenario.

```
Mibby messages bob.
/msg bob test
(Bob receives message.)

Mibby messages tom.
/msg tom test2
(Tom receives message.)

Bob messages mibby.
/msg mibby cake
(Mibby receives message.)

Mibby replies.
/r pie
(Tom receives message, not bob.)
```

It does not go to the person whose message you were last a part of.

**Comment by SupaHam** [comment]
Apologies for the inconvenience, this was my fault. Sadly the code is not in a good enough state to write proper test units. This issue has been fixed in the latest build of [EssentialsX](https://ci.drtshock.net/job/EssentialsX/218/).

**Comment by mibby** [comment]
@SupaHam The first issue is fixed (with players not being able to reply to new messages) but the second one is not, causing it to not emulate old behavior.

```
Mibby messages bob.
/msg bob test
(Bob receives message.)

Mibby messages tom.
/msg tom test2
(Tom receives message.)

Bob messages mibby.
/msg mibby cake
(Mibby receives message.)

Mibby replies.
/r pie
(Tom receives message, not bob.)
```

**Comment by SupaHam** [comment]
Not for me, can you double check.

I tried the exact scenario you described above, after Bob messages Mibby, Mibby's reply goes to Bob and not Tom.

**Comment by mibby** [comment]
With `last-message-reply-recipient: true` ? It doesn't seem to perform as intended for me. I'll take some screenshots if needed.

**Comment by SupaHam** [comment]
No, that was with `last-message-reply-recipient: false`. Which is the old behaviour, that is what you are talking about.

`last-message-reply-recipient: true` functions correctly, as we have resolved above.

**Comment by Kakifrucht** [comment]
last-message-reply-recipient: false does not seem to replicate the old behaviour properly. It replies to the last person that messaged you, and does not reply to the last person you have messaged (which was the old behaviour).

**Comment by mibby** [comment]
To clarify @Kakifrucht statement, old behavior replied to the last message that involved you. Not necessarily the last person you messaged or the last person that messaged you, but rather both - reply going to the person of the last message you received/sent.

**Comment by SupaHam** [comment]
So just to confirm, you're both saying the functionality missing in the old behaviour is that the sender doesn't have its reply-recipient set, even after it sends a message, correct?

**Comment by Kakifrucht** [comment]
If my english is not broken I do believe you got the point. Pretty much what @mibby said is what I was saying.

**Comment by SupaHam** [comment]
Please download [219](https://ci.drtshock.net/job/EssentialsX/219/) and confirm that-that is the expected behaviour.

**Comment by Kakifrucht** [comment]
Yeah, seems fixed now, thanks.

**Comment by mibby** [comment]
Confirmed for me as well. Seems to be working as intended now.

**Comment by Kakifrucht** [comment]
Little issue with old behaviour, when the person goes offline it prints both that he went offline and that he has messages disabled.

**Comment by SupaHam** [comment]
Fixed in fc8ee7f.
